File: week35/Leet1302.java
package me.algo.week35;

/**
 * Created by Bomi on 2021/01/16.
 * 문제 출처 : https://leetcode.com/problems/deepest-leaves-sum/submissions/
 *
 * Time Complexity : O(N)
 * Used Algorithm : DFS
 */

public class Leet1302 {
    private static int maxDepth;
    private static int sum;

    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public int deepestLeavesSum(TreeNode root) {
        deepestLeavesSum(root, 0);
        return sum;
    }

    private void deepestLeavesSum(TreeNode root, int currentDepth) {
        if (root == null) return;

        if (maxDepth == currentDepth) sum += root.val;
        if (maxDepth < currentDepth) {
            maxDepth = currentDepth;
            sum = root.val;
        }

        deepestLeavesSum(root.left, currentDepth + 1);
        deepestLeavesSum(root.right, currentDepth + 1);
    }
}

File: week52/sun/Leet1302.kt
package me.algo.week52.sun

/**
 * Created by Bomi on 2021/05/08.
 * 문제 출처 : https://leetcode.com/problems/deepest-leaves-sum/
 *
 * Time Complexity : O(N)
 * Used Algorithm : DFS
 * Used Data structure : Tree
 */

class Leet1302 {
    class TreeNode(var `val`: Int) {
        var left: TreeNode? = null
        var right: TreeNode? = null
    }

    private var deepestDepth = 0
    private var sum = 0
    fun deepestLeavesSum(root: TreeNode?): Int {
        deepestDepth = 0
        sum = 0
        deepestLeavesSum(root, deepestDepth)
        return sum
    }

    private fun deepestLeavesSum(root: TreeNode?, currentDepth: Int) {
        root?.let {
            if (deepestDepth == currentDepth) {
                sum += it.`val`
            }
            if (deepestDepth < currentDepth) {
                deepestDepth = currentDepth
                sum = it.`val`
            }
            deepestLeavesSum(it.left, currentDepth + 1)
            deepestLeavesSum(it.right, currentDepth + 1)
        }
    }
}